src/esp/gfx/PBRShader.cpp

```diff
@@ -0,0 +1,320 @@
+  CORRADE_ASSERT(flags_ & Flag::RoughnessTexture,
+                 "PBRShader::bindRoughnessTexture: the shader was not "
+                 "created with roughness texture enabled",
+                 *this);
+  if (lightCount_) {
+    texture.bind(TextureBindingPointIndex::roughness);
+  }
+  return *this;
+}
+
+PBRShader& PBRShader::bindMetallicTexture(Mn::GL::Texture2D& texture) {
+  CORRADE_ASSERT(flags_ & Flag::MetallicTexture,
+                 "PBRShader::bindMetallicTexture: the shader was not "
+                 "created with metallic texture enabled",
+                 *this);
+  if (lightCount_) {
+    texture.bind(TextureBindingPointIndex::metallic);
+  }
+  return *this;
+}
+
+PBRShader& PBRShader::bindNormalTexture(Mn::GL::Texture2D& texture) {
+  CORRADE_ASSERT(flags_ & Flag::NormalTexture,
+                 "PBRShader::bindNormalTexture: the shader was not "
+                 "created with normal texture enabled",
+                 *this);
+  if (lightCount_) {
+    texture.bind(TextureBindingPointIndex::normal);
+  }
+  return *this;
+}
+
+PBRShader& PBRShader::setMVPMatrix(const Mn::Matrix4& matrix) {
+  setUniform(mvpMatrixUniform_, matrix);
+  return *this;
+}
+
+PBRShader& PBRShader::setNormalMatrix(const Mn::Matrix3x3& matrix) {
+  setUniform(normalMatrixUniform_, matrix);
+  return *this;
+}
+
+PBRShader& PBRShader::setTransformationMatrix(const Mn::Matrix4& matrix) {
+  setUniform(modelviewMatrixUniform_, matrix);
+  return *this;
+}
+
+PBRShader& PBRShader::setObjectId(unsigned int objectId) {
+  if (flags_ & Flag::ObjectId) {
+    setUniform(objectIdUniform_, objectId);
+  }
+  return *this;
+}
+
+PBRShader& PBRShader::setBaseColor(const Mn::Color4& color) {
+  if (lightCount_) {
+    setUniform(baseColorUniform_, color);
+  }
+  return *this;
+}
+
+PBRShader& PBRShader::setRoughness(float roughness) {
+  if (lightCount_) {
+    setUniform(roughnessUniform_, roughness);
+  }
+  return *this;
+}
+
+PBRShader& PBRShader::setMetallic(float metallic) {
+  if (lightCount_) {
+    setUniform(metallicUniform_, metallic);
+  }
+  return *this;
+}
+
+PBRShader& PBRShader::bindTextures(Mn::GL::Texture2D* albedo,
+                                   Mn::GL::Texture2D* roughness,
+                                   Mn::GL::Texture2D* metallic,
+                                   Mn::GL::Texture2D* normal) {
+  if (albedo) {
+    bindAlbedoTexture(*albedo);
+  }
+  if (roughness) {
+    bindRoughnessTexture(*roughness);
+  }
+  if (metallic) {
+    bindMetallicTexture(*metallic);
+  }
+  if (normal) {
+    bindNormalTexture(*normal);
+  }
+  return *this;
+}
+
+PBRShader& PBRShader::setLightPosition(unsigned int lightIndex,
+                                       const Mn::Vector3& pos) {
+  CORRADE_ASSERT(
+      lightIndex < lightCount_,
+      "PBRShader::setLightPosition: lightIndex" << lightIndex << "is illegal.",
+      *this);
+
+  setUniform(lightDirectionsUniform_[lightIndex], Mn::Vector4{pos, 1.0});
+  return *this;
+}
```

Comment:
Any chance all positions could be specified together in a single call (and colors / ranges in another), like [it's done in Phong](https://doc.magnum.graphics/magnum/classMagnum_1_1Shaders_1_1Phong.html#aa25f246ff11ce2fc5d069cd34fd1ea6a)? Because calling `setUnform()` 30 times for 10 lights instead of just three times will have a pretty large impact on API overhead, *especially* on the web.

If you do it like I do in Phong, it'll also have the nice property of all array elements having consecutive uniform indices, which means you wouldn't need to store uniform locations in a vector.

Alternatively throw all this away and go with UBOs instead -- all targets supported by Habitat have those.

Reply:
> means that you will have to compute the "location" precisely and put such "flexible length" arrays always in the very end

I do that so all non-array uniforms can have explicit location, yes, but since you can't use that here, it doesn't matter. It's enough to query the location of the first element and then assume all other elements will occupy the locations after, that always holds.